Refactor funcol for readability and dynamo tracing

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104387

Move eager kernel impls to separate file, which is eaiser to read
(since users may be confused about 2 versions of each kernel in the same file)
and easier to set a dynamo policy to trace only the first file currently.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng

wconstab committed:
Refactor funcol for readability and dynamo tracing

Move eager kernel impls to separate file, which is eaiser to read
(since users may be confused about 2 versions of each kernel in the same file)
and easier to set a dynamo policy to trace only the first file currently.

[ghstack-poisoned]

wconstab commented:
cc @kumpera @wanchaol

wconstab committed:
Update on "Refactor funcol for readability and dynamo tracing"


Move eager kernel impls to separate file, which is eaiser to read
(since users may be confused about 2 versions of each kernel in the same file)
and easier to set a dynamo policy to trace only the first file currently.

[ghstack-poisoned]

wanchaol commented on torch/distributed/_functional_collectives.py:
```diff
@@ -1,12 +1,11 @@
 from typing import Tuple, Union, List, cast, TYPE_CHECKING
 from torch.utils._pytree import tree_map_only
-
+from ._functional_collectives_impl import _all_reduce, _register_wrapper_tensor
```
Curious why we only need to import `_all_reduce` impl? I thought we need to import all the collective kernel impls.

wconstab replied:
technically we don't need to import any.

I import it so that i can query it for its '__module__' and pass that module obj to the registration code, which will search/import all the ops by name from that module.

I could have used the `_register_tensor_wrapper` obj to get that module but for some reason i felt its cleaner to use the module that has one of the collectives in it.

wanchaol replied:
I see, in this case, can we just import the module instead? i.e. `import ._functional_collectives_impl as funcol_impl`, we can then pass the `funcol_impl` module for registration?

wconstab committed:
Update on "Refactor funcol for readability and dynamo tracing"


Move eager kernel impls to separate file, which is eaiser to read
(since users may be confused about 2 versions of each kernel in the same file)
and easier to set a dynamo policy to trace only the first file currently.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

wconstab committed:
Update on "Refactor funcol for readability and dynamo tracing"


Move eager kernel impls to separate file, which is eaiser to read
(since users may be confused about 2 versions of each kernel in the same file)
and easier to set a dynamo policy to trace only the first file currently.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

wanchaol reviewed (approved):
stamp to unblock, looks like the import statement did not get updated?

wconstab committed:
Update on "Refactor funcol for readability and dynamo tracing"


Move eager kernel impls to separate file, which is eaiser to read
(since users may be confused about 2 versions of each kernel in the same file)
and easier to set a dynamo policy to trace only the first file currently.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

kumpera commented on torch/distributed/_functional_collectives_impl.py:
```diff
@@ -0,0 +1,246 @@
+
+"""
+Moved eager kernel implementations to a separate file partly for readability and partly as it is currently
+easier in dynamo to set tracing policy on a file-by-file level.
```
This is interesting, are there some guidelines of what should go on this file and what should go in _functional_collectives.py?

kumpera reviewed (approved):
LGTM.

This should work fine internally cuz we're not touching the registration code.

wconstab commented:
@pytorchbot merge -i

wconstab commented:
@pytorchbot rebase

wconstab committed:
Update on "Refactor funcol for readability and dynamo tracing"


Move eager kernel impls to separate file, which is eaiser to read
(since users may be confused about 2 versions of each kernel in the same file)
and easier to set a dynamo policy to trace only the first file currently.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab committed:
Update on "Refactor funcol for readability and dynamo tracing"


Move eager kernel impls to separate file, which is eaiser to read
(since users may be confused about 2 versions of each kernel in the same file)
and easier to set a dynamo policy to trace only the first file currently.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng

[ghstack-poisoned]

wconstab commented:
@pytorchhbot merge -i

wconstab commented:
@pytorchbot merge -i
